### PR TITLE
Integrate BREEDS helpers and documentation

### PR DESCRIPTION
**Pull request by ShibaniSanturkar**
Modify helpers for the BREEDS dataset and add documentation.

### CODE REVIEW

File: docs/example_usage/breeds_datasets.rst [hunk 1]
```diff
@@ -0,0 +1,288 @@
+Creating BREEDS subpopulation shift benchmarks
+===============================================
+
+In this document, we will discuss how to create BREEDS datasets, used in 
+the `code <https://github.com/MadryLab/BREEDS_benchmarks>`_ `release
+<https://github.com/MadryLab/BREEDS_benchmarks>`_ based on our paper [STM20]_.
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:6]: Remove "our paper" IMO

File: docs/example_usage/breeds_datasets.rst [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+At a high-level, the procedure to construct BREEDS datasets is as follows:
+
+1. Use the dataset's class hierarchy to define `superclasses` by grouping
+   together semantically similar dataset classes (e.g., define the `dog`` superclass 
+   by aggregating all the dog breed classes). 
+2. Design the classification task to be between different superclasses, so that
+   we have easy access to the individual subpopulations (i.e., the original 
+   dataset classes). Then, for each superclass, we will partition the subpopulations 
+   into the source (used for model training) and target domain without overlap
+   (used for evaluation).
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:27]: Suggested edit:

1. Use the ImageNet hierarchy to group together similar classes into "superclasses," e.g., we can make a "dog" superclass by ...
2. Partition the subclasses of each superclass into "training breeds" and "testing breeds." The classification task will be differentiating between the superclasses, but at train time we will see only the "training breeds" from each superclass, and at testing time we will see only the testing breeds.

File: docs/example_usage/breeds_datasets.rst [hunk 3]
```diff
@@ -0,0 +1,288 @@
+Creating BREEDS subpopulation shift benchmarks
+===============================================
+
+In this document, we will discuss how to create BREEDS datasets, used in 
+the `code <https://github.com/MadryLab/BREEDS_benchmarks>`_ `release
+<https://github.com/MadryLab/BREEDS_benchmarks>`_ based on our paper [STM20]_. 
+These datasets can be used to study model robustness to subpopulation 
+shift---a specific form of distribution shift wherein the subpopulations of data 
+(e.g., animal breeds) present in the training set (source domain) are 
+entirely different from the test set (target domain).
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:10]: Give an example of a dog vs cat classifier: e.g. "On the train set we are classifying between tabby cats and labradors, and in the test set we are classifying between Egyptian cats and golden retrievers." It's kind of hard right now to tell exactly what the dataset is early on in the docs

File: docs/example_usage/breeds_datasets.rst [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+shift---a specific form of distribution shift wherein the subpopulations of data 
+(e.g., animal breeds) present in the training set (source domain) are 
+entirely different from the test set (target domain). 
+You can create BREEDS datasets from any existing dataset that has 
+an explicit class hierarchy describing the semantic relationship 
+between pairs of classes (e.g., ImageNet, OpenImages). 
+In this walkthrough, we will do so using ImageNet and the corresponding
+class hierarchy from [STM20]_ (modification of WordNet to be better-suited
+for the object recognition task).
+
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:17]: Is there a summarizing picture of the class hierarchy

File: docs/example_usage/breeds_datasets.rst [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+  For ImageNet, you can find the relevant files in the 
+  `imagenet_class_hierarchy/modified` subfolder of our
+  `release <https://github.com/MadryLab/BREEDS_benchmarks>`_. 
+
+Part 1: Browsing through the Class Hierarchy
+''''''''''''''''''''''''''''''''''''''''''''
+The :class:`~robustness.tools.breeds_helpers.ClassHierarchy` class allows
+us to probe the class hierarchy. The class hierarchy has a graph structure,
+where a child node is a subclass (an instance) of the parent node and the
+original dataset classes are the leaves. We use a top-down approach to obtain
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:53]: Rest of this paragraph seems like a footnote or a parenthetical, the main thing is "there is a tree that describes all of the ImageNet classes"

File: docs/example_usage/breeds_datasets.rst [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+
+.. raw:: html
+
+   <i class="fa fa-play"></i> &nbsp;&nbsp; <a
+   href="https://github.com/MadryLab/BREEDS-Benchmarks/blob/master/Constructing%20BREEDS%20datasets.ipynb">Download
+   a Jupyter notebook</a> containing all the code from this walkthrough! <br />
+   <br />
+
+Requirements/Setup
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:37]: Would be nice to have a setup_breeds(root=SOME_DIR) function (noted this below too) that downloads all of this stuff and puts it in root, that way they don't have to clone the other repo.

File: docs/example_usage/breeds_datasets.rst [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+.. code-block:: python
+
+   from robustness.tools.breeds_helpers import ClassHierarchy
+   hier = ClassHierarchy(info_dir)
+
+
+Here, :samp:`info_dir` should be the path to a folder which contains
+information about the class hierarchy 
+(e.g., ``imagenet_class_hierarchy/modified``). In general, :samp:`info_dir`
+must contain the following files:
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:70]: Just say that info_dir is the repo you downloaded in the previous step, this part is confusing

File: docs/example_usage/breeds_datasets.rst [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+                    superclasses, 
+                    hier.LEAF_NUM_TO_NAME)
+
+Finally, for the source and target domains, we can create datasets
+and their corresponding loaders:
+
+.. code-block:: python
+  from robustness.datasets import DATASETS
+  # For the source domain
+  dataset_source = DATASETS['custom_imagenet'](data_dir, subclass_tuple[0])
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:230]: should just do (although as noted above I think the name of subclass_tuple should change):
```python
from robustness import datasets
train_subclasses, test_subclasses = subclass_tuple
dataset_source = datasets.CustomImageNet(data_dir, train_subclasses)
```

File: docs/example_usage/breeds_datasets.rst [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+  train_loader_source, val_loader_source = dataset_source.make_loaders(num_workers, 
+                                                                     batch_size)
+  # For the target domain                                                                     
+  dataset_target = DATASETS['custom_imagenet'](data_dir, subclass_tuple[1])
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:234]: Same as above, avoid using `DATASETS`

File: docs/example_usage/breeds_datasets.rst [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+.. code-block:: python
+
+   ret = DG.get_superclasses(level=2, 
+                          Nsubclasses=None, 
+                          split="rand", 
+                          ancestor=None, 
+                          balanced=True)
+  subclass_ranges, label_map, subclass_tuple, superclasses, _ = ret                                    
+
+This method returns:
```
Comment (andrewilyas) [inline docs/example_usage/breeds_datasets.rst:191]: - List the return values below in the same order they are returned
- What is the `_` in the return value? 
- What is the point of `subclass_ranges`, isn't it just a straightforward transformation of `subclass_tuple`
- Rename `subclass_tuple` to like `subclass_splits` or something more descriptive IMO

File: docs/example_usage/breeds_datasets.rst [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+Part 3: Loading in-built BREEDS Datasets
+''''''''''''''''''''''''''''''''''''''''
+
+Alternatively, we can directly use one of the datasets from our paper 
+[STM20]_---namely ``Entity13``, ``Entity30``, ``Living17`` 
+and ``Nonliving26``. Loading any of these datasets is relatively simple:
+
+.. code-block:: python
+
+  from robustness.tools.breeds_helpers import Living17
```
[the remark this answers is not in the document]
Comment (andrewilyas) [reply docs/example_usage/breeds_datasets.rst:282]: Ok reading the rest of the tutorial I can see why this is, but IMO the notation should be a bit different, these should be functions called make_living_17 or make_mixed_10 or whatever, to make it clear that these are helper *functions* and not actual dataset classes.

File: robustness/loaders.py [hunk 12]
```diff
@@ -60,8 +60,10 @@ def make_loaders(workers, batch_size, transforms, data_path, data_aug=True,
     else:
         if not only_val:
             train_set = custom_class(root=data_path, train=True, 
+                                        label_mapping=label_mapping,
                                         download=True, transform=transform_train)
         test_set = custom_class(root=data_path, train=False, 
+                                    label_mapping=label_mapping,
```
Comment (andrewilyas) [inline robustness/loaders.py:66]: Does this break loading other dataset with custom classes? It looks like it will which would be bad. If it does let me know because I already have a fix in mind.

File: robustness/tools/breeds_helpers.py [hunk 13]
```diff
@@ -10,35 +11,48 @@ class ClassHierarchy():
...
         """
         Args:
             info_dir (str) : Path to hierarchy information files. Contains a 
                 "class_hierarchy.txt" file with one edge per line, a
                 "node_names.txt" mapping nodes to names, and "class_info.json".
         """
         
-        with open(f'{info_dir}/class_info.json', 'r') as f:
+        # Details about dataset class (sleaves) names, IDS
+        with open(f'{info_dir}/dataset_class_info.json', 'r') as f:
```
Comment (andrewilyas) [inline robustness/tools/breeds_helpers.py:23]: Can we have a user-friendly error (maybe with a link to the docs and the breeds repo) if the modified hierarchy file is not found?

File: docs/example_usage/breeds_datasets.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (ShibaniSanturkar) [reply docs/example_usage/breeds_datasets.rst:282]: Fixed.

File: docs/example_usage/breeds_datasets.rst (hunk 10 above)
Comment (ShibaniSanturkar) [reply docs/example_usage/breeds_datasets.rst:191]: Fixed and reduced returns.

File: docs/example_usage/breeds_datasets.rst (hunk 6 above)
Comment (ShibaniSanturkar) [reply docs/example_usage/breeds_datasets.rst:37]: Done.
